### PR TITLE
feat: add creator course redemption codes

### PR DESCRIPTION
**Pull request by QinyingShi**
## Summary
- Add a creator-side redemption code tab under Admin Orders, with create/list/search support for the creator's own published courses.
- Add creator-owned redemption code detail, edit, enable/disable, usage records, sub-code list, and sub-code export support.
- Add creator columns to operator promotion coupon and campaign tables, and reuse shared admin row actions/dialogs where possible.

## Tests
- `cd src/api && pytest tests/service/promo/test_admin_promotions.py tests/service/billing/test_legacy_boundary_contracts.py -q`
- `cd src/cook-web && npm test -- src/app/admin/orders/CreatorRedemptionCodesTab.test.tsx src/app/admin/orders/CreatorRedemptionCodeDialog.test.tsx src/app/admin/orders/page.test.tsx src/app/admin/operations/promotions/page.test.tsx --runInBand`
- `cd src/cook-web && npm run type-check`
- `cd src/cook-web && npm run lint -- --file src/app/admin/orders/CreatorRedemptionCodesTab.tsx --file src/app/admin/orders/CreatorRedemptionCodesTab.test.tsx --file src/app/admin/orders/CreatorRedemptionCodeDialog.tsx --file src/app/admin/orders/CreatorRedemptionCodeDialog.test.tsx --file src/app/admin/orders/page.tsx --file src/app/admin/orders/page.test.tsx --file src/app/admin/operations/promotions/PromotionRecordDialogs.tsx --file src/app/admin/operations/promotions/page.test.tsx`
- `python scripts/check_architecture_boundaries.py`
- `pre-commit run -a`

## Notes
- Local Docker API/nginx was restarted and `/api/runtime-config` returned 200.
- Follow-up performance items are recorded in the local operator optimization plan and are not included in this PR.

## Summary by CodeRabbit

* **New Features**
  * Creators can create, list, view, update, enable/disable, and export course-scoped redemption codes with filtering, pagination, usages and sub-code lists; strict input validation and date parsing applied.
  * Admin UI: new Redemption Codes tab, create/edit dialog, creator column in promotions tables, row actions, export codes flow, and URL-driven tab navigation.
  * DTOs/UI now surface creator name/ID for coupons and campaigns.

* **Tests**
  * Expanded tests for creator redemption-code flows, ownership enforcement, dialogs, export, concurrency and edge cases.

* **i18n**
  * Added/localized strings for redemption code flows (en/fr/zh).

[Review Change Stack](https://app.coderabbit.ai/change-stack/ai-shifu/ai-shifu/pull/1851?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/api/flaskr/route/order.py [hunk 1]
```diff
@@ -622,6 +668,131 @@
...
+        if not isinstance(payload, dict):
+            raise_param_error("payload")
+        return make_common_response(
+            update_creator_course_redemption_coupon_status(
+                app,
+                request.user.user_id,
+                coupon_bid,
+                payload.get("enabled"),
+            )
+        )
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/route/order.py]: Fixed. The status endpoint now rejects non-object JSON payloads and parses enabled into a strict boolean before calling the service, so raw request values are not reflected in the response path.

File: src/i18n/fr-FR/modules/order.json [hunk 2]
```diff
@@ -118,6 +118,26 @@
...
   },
+  "redemptionCodes": {
+    "action": "Créer un code",
+    "courseLabel": "Cours",
+    "coursePlaceholder": "Sélectionner un cours publié",
+    "courseRequired": "Le cours est requis",
+    "createFailed": "Échec de la création du code",
+    "createSuccess": "Créé avec succès",
+    "dialogTitle": "Créer un code de réduction",
+    "editDialogTitle": "Modifier le code de redemption",
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/i18n/fr-FR/modules/order.json]: Fixed. Updated the French edit title to use “code de réduction” consistently.

File: src/i18n/fr-FR/modules/order.json [hunk 3]
```diff
@@ -118,6 +118,26 @@
...
+    "emptyCourses": "Aucun cours publié",
+    "emptyList": "Aucun code",
+    "loadCoursesFailed": "Échec du chargement des cours. Veuillez réessayer plus tard.",
+    "loadFailed": "Échec du chargement des codes. Veuillez actualiser et réessayer.",
+    "submitting": "Création...",
+    "totalCount": "Codes: {count}",
+    "updateSuccess": "Enregistre avec succes",
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/i18n/fr-FR/modules/order.json]: Fixed. Updated the French success message to “Enregistré avec succès”.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodeDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+            if (course?.bid && !seen.has(course.bid)) {
+              seen.add(course.bid);
+              collected.push(course);
+            }
+          });
+          if (pageItems.length < pageSize) {
+            break;
+          }
+          pageIndex += 1;
+        }
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodeDialog.tsx]: Fixed. Added a maximum page cap for loading published courses and a localized truncation message, with test coverage.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodeDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+    () => courses.filter(course => Boolean(String(course.bid || '').trim())),
+    [courses],
+  );
+
+  const handleSubmit = async () => {
+    const normalizedName = form.name.trim();
+    const normalizedCode = form.code.trim();
+    const normalizedQuantity = form.total_count.trim();
+    const normalizedCourseId = form.shifu_bid.trim();
+    const normalizedValue = form.value.trim();
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodeDialog.tsx]: Fixed. The submit handler now safely normalizes optional form fields before trimming.

File: src/api/flaskr/route/order.py [hunk 6]
```diff
@@ -31,6 +42,41 @@ def _require_creator():
...
+        raise_param_error("datetime format invalid")
+
+    def _parse_admin_pagination():
+        page_index = request.args.get("page_index", 1)
+        page_size = request.args.get("page_size", 20)
+        try:
+            page_index = int(page_index)
+            page_size = int(page_size)
+        except ValueError:
+            raise_param_error("page_index or page_size is not a number")
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/route/order.py]: Fixed. Pagination parsing now validates page_index and page_size separately and raises field-specific parameter errors.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodeDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+
+  useEffect(() => {
+    if (!open) {
+      return undefined;
+    }
+
+    let canceled = false;
+    const loadCourses = async () => {
+      setCoursesLoading(true);
+      setCoursesError('');
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodeDialog.tsx]: Fixed. The course error reset now uses a defensive functional state updater.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx [hunk 8]
```diff
@@ -0,0 +1,807 @@
...
+
+  useEffect(() => {
+    filtersRef.current = filters;
+  }, [filters]);
+
+  const fetchCodes = useCallback(
+    async (targetPage: number, nextFilters?: RedemptionCodeFilters) => {
+      const resolvedFilters = nextFilters ?? filtersRef.current;
+      setLoading(true);
+      setError(null);
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx]: Fixed. Error reset in the redemption-code list now uses a defensive functional state updater.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx [hunk 9]
```diff
@@ -0,0 +1,807 @@
...
+      await api.updateCreatorCourseRedemptionCodeStatus({
+        coupon_bid: statusTarget.coupon_bid,
+        enabled: enabling,
+      });
+      showDefaultToast(
+        enabling
+          ? tPromotion('messages.couponEnabledSuccess')
+          : tPromotion('messages.couponDisabledSuccess'),
+      );
+      setStatusTarget(null);
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx]: Fixed. Status target reset now uses a defensive functional state updater.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx [hunk 10]
```diff
@@ -0,0 +1,807 @@
...
+        open={codesDialogOpen}
+        onOpenChange={setCodesDialogOpen}
+        couponBid={selectedCouponBid}
+        couponName={selectedCouponName}
+        fetchCodesApi={fetchCreatorSubCodes}
+      />
+      <CreatorRedemptionCodeDialog
+        open={Boolean(editingCoupon)}
+        onOpenChange={open => {
+          if (!open) {
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx:795]: Fixed. Editing coupon reset now uses a defensive functional state updater.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx [hunk 11]
```diff
@@ -0,0 +1,807 @@
...
+            ? {
+                entityType: 'coupon',
+                enabling: statusTarget.computed_status === 'inactive',
+                item: statusTarget,
+              }
+            : null
+        }
+        submitting={statusSubmitting}
+        onOpenChange={open => {
+          if (!open && !statusSubmitting) {
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx:814]: Fixed. Status target reset on dialog close now uses a defensive functional state updater.

File: src/api/flaskr/route/order.py [hunk 12]
```diff
@@ -31,6 +42,41 @@ def _require_creator():
...
+        if page_index < 1 or page_size < 1:
+            raise_param_error("page_index or page_size is less than 1")
+        return page_index, page_size
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/route/order.py]: Fixed. Removed raw English parameter-error strings and now raises field-specific parameter errors.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx [hunk 13]
```diff
@@ -0,0 +1,807 @@
...
+      } catch (err) {
+        setError({
+          message:
+            err instanceof Error
+              ? err.message
+              : t('module.order.redemptionCodes.loadFailed'),
+        });
+      } finally {
+        setLoading(false);
+      }
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx:191]: Fixed. Added request sequencing to ignore stale list responses, with regression coverage for out-of-order responses.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx [hunk 14]
```diff
@@ -0,0 +1,807 @@
...
+          errorMessage={error.message}
+          errorCode={0}
+        />
+      ) : null}
+
+      <AdminTableShell
+        loading={loading}
+        isEmpty={items.length === 0}
+        emptyContent={t('module.order.redemptionCodes.emptyList')}
+        emptyColSpan={11}
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx:581]: Fixed. The empty state is now hidden while loading or when a fetch error is present, with regression coverage.

File: src/api/flaskr/route/order.py [hunk 15]
```diff
@@ -622,6 +668,131 @@ def admin_import_activation():
...
+            )
+        )
+
+    @app.route(path_prefix + "/admin/orders/redemption-codes", methods=["POST"])
+    def admin_create_creator_redemption_code():
+        """Create a course redemption code for the current creator's published course."""
+        _require_creator()
+        payload = request.get_json(silent=True) or {}
+        if not isinstance(payload, dict):
+            raise_param_error("payload")
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/route/order.py]: Fixed. The create/update/status handlers now reject missing, invalid, and non-object JSON payloads before calling the service.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx [hunk 16]
```diff
@@ -0,0 +1,818 @@
...
+      } catch (err) {
+        if (requestId !== fetchRequestIdRef.current) {
+          return;
+        }
+        setError({
+          message:
+            err instanceof Error
+              ? err.message
+              : t('module.order.redemptionCodes.loadFailed'),
+        });
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodesTab.tsx:186]: Fixed. On current-request load failure, the list now clears stale rows, pagination, and totals so failed reloads cannot show previous results. Added regression coverage for this case.

File: src/cook-web/src/app/admin/orders/CreatorRedemptionCodeDialog.tsx [hunk 17]
```diff
@@ -0,0 +1,563 @@
...
+              </SelectTrigger>
+              <SelectContent className='max-h-60'>
+                {coursesLoading ? (
+                  <div className='flex items-center justify-center py-3'>
+                    <Loading className='h-5 w-5' />
+                  </div>
+                ) : coursesError ? (
+                  <div className='px-2 py-3 text-xs text-destructive'>
+                    {coursesError}
+                  </div>
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/orders/CreatorRedemptionCodeDialog.tsx:486]: Fixed. The course-limit path now uses a warning instead of a blocking error, so collected course options remain visible and selectable. Added regression coverage to keep the loaded options visible.
